packages/protocol/contracts/ACE/noteRegistry/Manager.sol

```diff
@@ -0,0 +1,334 @@
+            // to their byte positions in _factoryId.
+            // i.e. (epoch * 0x10000) = and(_factoryId, 0xffff0000)
+            // and  (cryptoSystem * 0x100) = and(_factoryId, 0xff00)
+            // and  (assetType) = and(_factoryId, 0xff)
+
+            // Putting this all together. The storage slot offset from '_factoryId' is...
+            // (_factoryId & 0xffff0000) + (_factoryId & 0xff00) + (_factoryId & 0xff)
+            // i.e. the storage slot offset IS the value of _factoryId
+            factoryAddress := sload(add(_factoryId, factories_slot))
+
+            isFactoryDisabled :=
+                shr(
+                    shl(0x03, and(_factoryId, 0x1f)),
+                    sload(add(shr(5, _factoryId), disabledFactories_slot))
+                )
+            queryInvalid := or(iszero(factoryAddress), isFactoryDisabled)
+        }
+
+        // wrap both require checks in a single if test. This means the happy path only has 1 conditional jump
+        if (queryInvalid) {
+            require(factoryAddress != address(0x0), "expected the factory address to exist");
+            require(isFactoryDisabled == false, "expected the factory address to not be disabled");
+        }
+    }
+
+    /**
+    * @dev Call transferFrom on a linked ERC20 token. Used in cases where the ACE's mint
+    * function is called but the token balance of the note registry in question is
+    * insufficient
+    *
+    * @param _value the value to be transferred
+    */
+    function supplementTokens(uint256 _value) external {
+        NoteRegistryBehaviour registry = registries[msg.sender];
+        require(address(registry) != address(0x0), "note registry does not exist");
+        (
+            address linkedToken,
+            uint256 scalingFactor,
+            ,,,
+            bool canConvert,
+            bool canAdjustSupply
+        ) = registry.getRegistry();
+        require(canConvert == true, "note registry does not have conversion rights");
+        require(canAdjustSupply == true, "note registry does not have mint and burn rights");
+        IERC20(linkedToken).transferFrom(msg.sender, address(this), _value.mul(scalingFactor));
+        registry.supplementTokens(_value);
+    }
+
+    /**
+    * @dev Query the ACE for a previously validated proof
+    * @notice This is a virtual function, that must be overwritten by the contract that inherits from NoteRegistr
+    *
+    * @param _proof - unique identifier for the proof in question and being validated
+    * @param _proofHash - keccak256 hash of a bytes proofOutput argument. Used to identify the proof in question
+    * @param _sender - address of the entity that originally validated the proof
+    * @return boolean - true if the proof has previously been validated, false if not
+    */
+    function validateProofByHash(uint24 _proof, bytes32 _proofHash, address _sender) public view returns (bool);
+
+    function createNoteRegistry(
```

Comment:
I decided against this so that the interface the current ZkAssets use doesn't change. Both methods are still available to users, depending on arguments passed.

Reply:
Fair enough